Add a NEST documentation style guide

jessica-mitchell opened the pull request:
We've been discussing having a style guide for quite some time with the aim to help reduce inconsistencies in documentation writing.
This PR introduces a style guide inspired from several other style guides around the interwebs.

It includes several sections regarding English language as well as formatting and markup notation.

**Please note:**

There is a lot that still needs to be discussed here! @jougs @heplesser @terhorstd I'm sure you all have thoughts on this topic.

- Any information that is missing? Anything too excessive?

- Do we want to use an external style guide (like Microsoft's or Google's style guide) as a reference, in case people run into questions we do not address in our style guide? 

- Some of the choices like formatting marks made here are based on what we do _usually_ or what has been discussed, but obviously we can make changes as we see fit. We just need to make sure that readability is improved with the choices we make not impeded.

Thanks to @sarakonradi for contributing to this as well.

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+Admonitions
+^^^^^^^^^^^
+
+Use admonitions to draw readers attention to a particular point. Possible
+admonitions types include
```
```suggestion
Use admonitions to draw the reader's attention to a particular point. Possible
admonition types include
```

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+Referencing
+^^^^^^^^^^^
+
+For referencing reStructuredText files in repository, use
```
```suggestion
For referencing reStructuredText files within the documentation, use
```

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+::
+
+   :doc:`custom name <path/file>`
```
```suggestion
   :doc:`custom label <path/file>`
```

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+   :doc:`file`
+
+   :doc:`custom name <file>`
```
```suggestion
   :doc:`custom label <file>`
```

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+For section headings, you can use either of the two examples below. But to
+reference figures or arbitrary places in a file, you must include a custom
+name (second example) in the reference for it to work.
```
```suggestion
label (second example) in the reference for it to work.
```

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+::
+
+    :ref:`custom name <ref_name>`
```
```suggestion
    :ref:`custom label <ref_name>`
```

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+    :ref:`ref_name`
+
+    :ref:`custom name <ref_name>`
```
```suggestion
    :ref:`custom label <ref_name>`
```

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+        .. _ref_name:
+
+        secton_header
```
```suggestion
        section_header
```

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+::
+
+   :py:func:`.Create`
```
Just out of interest: do you have plans to add these automatically by scanning all double-backtick-delimited code and matching against the list of known things (PyNEST functions, models, ...)? I think this would be really awesome and make the process of documentation writing much more pleasant.

heplesser replied:
Indeed. 

Should this then result in a general recommendation that wherever a part (module, class, method, function) from the PyNEST API is mentioned in the documentation, one should use 
```
:py:func:`.Create`
```
instead of 
```
``Create``
```
And how would that then integrate with `nest.Create("iaf_psc_alpha")`?

jessica-mitchell replied:
@jougs @heplesser The PR to add links is in draft mode  #2010.

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+   :py:func:`.Create`
+
+
+To link terms to the glossary page use the HoverXTooltip role from Mahdi Enan (INM-6)
```
Can we link to somewhere from the name? Maybe a GitHub repository or such?

heplesser replied:
Is this role available from a general repository for Sphinx/RST resources? Does it need to be installed? If it is not available from general source, I think we should include the necessary code in the NEST repo to avoid dependencies on personal repos.

clinssen replied:
Or under the NEST or INM-6 account/organisation, so it can be used by NESTML as well!

jessica-mitchell replied:
@jougs @heplesser @clinssen Mahdi developed this Sphinx extension in the INM-6 repo: https://github.com/INM-6/HoverXTooltip, It was put there as recommended by @terhorstd, since he knows more about the regulations and best practices to this kind of thing. I can add a link to the repo if that is appropriate.

Until this Hovertooltip is discussed and merged, same with the links to functions and intersphinx, I'd want to remove these  from the styleguide. I put them here now in case their PRs are ready while this one is still open.

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+One word subheadings are acceptable, if the section is short and the meaning is clear.
+
+Use sentence case for headings and subheadings, i.e., begin with an uppercase
+letter but with all other words in lower case (except proper nouns).
```
I wonder if we should enforce this in displaying by using a CSS `text-transform` on the resulting `<hX>`tags. Or would you rather enforce this by having a linting check during the static code check phase in the CI?

heplesser replied:
There will always be exceptions, e.g. when `NEST` is part of a heading ...

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+   Example:
+
+   We have over 5,000 connections.
+   The number of connections is ``x = 5000``
```
```suggestion
   The number of connections is ``x = 5001``
```
> only to justify the word *over* in the line above ;-)

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+   The number of connections is ``x = 5000``
+
+Make sure you use the correct unit (e.g., millivolts for voltage) and the
+unit's syntax (`V_m`).
```
```suggestion
unit's symbol (`V_m`).
```
Without this change, the sentence does not compute for me.

heplesser replied:
I am confused here. `V_m` is not a unit, it is a parameter name. Could you give a complete example (complete sentence)?

clinssen replied:
Is a reference to nest::names appropriate here?

jougs replied:
@heplesser: you're right. What about so:

```suggestion
unit's symbol (`mV`).
```

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+If bullet/ordered list text is an incomplete sentence, do not end with period.
+
+If bullet/ordered list belongs to a sentence, use commas to separate each item
+with the second last time including `and` at end.
```
In this case, I usually use a period after the last item to indicate the end of the list sentence. I think we should mention if we want this or not.

jessica-mitchell replied:
added

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Spell out acronyms on first appearance on each page or article it appears.
+For example: Random number generator (rng)
```
```suggestion
For example: Random number generator (RNG)
```
Or should abbreviations really be lower case?

heplesser replied:
Link to glossary as well where that makes sense.

jougs commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+Capitalize first word in bullet.
+
+Capitalize proper nouns and follow company policy in naming conventions
+(e.g., macOS, LaTeX).
```
```suggestion
(e.g., macOS, LaTeX, Python, NumPy, NEST).
```

jougs reviewed (commented):
Many many thanks for this :heart:

I have added some concrete suggestions and some discussion points and questions inline.

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+   reStructuredText is sensitive to indentation. A single newline should always
+   be between any text and headings, tables, reference labels, or directives.
+   If the output format seems incorrect, double check the spaces and newlines
+   of the text.
```
And add a link to the Doc Workflow document about how to build the documentation so you can check at all. We have no control over how people arrive at this page, so this page may be the first they read when looking for documentation about NEST.

jessica-mitchell replied:
link added

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+           print(thing.x)
+
+For other code styles use the code-block directive (you can also use this one
+for python)
```
If I  wanted to use "this" for Python, it would presumably have to be

``
.. code-block:: python
``

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+Equations
+^^^^^^^^^
+
+For equations, use LaTeX markup:
```
Add information about inline math as well.

jessica-mitchell replied:
This has been reworded

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+    .. math::
+
+            f(x) = \int_{-\infty}^{\infty} \hat f(\xi)\ e^{2 \pi i x \xi}\,d\xi,
```
I am not sure I agree with the LaTeX typesetting here. If I remember my Knuth correctly, it should be
```suggestion
            f(x) = \int_{-\infty}^{\infty} \hat{f}(\xi) \mathrm{e}^{2 \pi i x \xi} \mathrm{d}\xi \;.
```
The logic is as follows:
- LaTeX handles whitespace within math expressions automatically
- Since e and d are no variables, they should be set in roman
- In the example here, the equation is at the end of a sentence and should thus end with a fullstop. The fullstop is set off with a slightly wider space from the equation.

The same changes apply to the "rendered" line below.

jessica-mitchell replied:
I copied this from somewhere (Wikipedia?) as my LaTeX equation knowledge is virtually non-existent; so if it renders, I interpret it to be correct.
Do we need some sort of reference here to show how to properly format LaTeX?

heplesser replied:
I modified my suggestion above. There are different schools of thought especially on whether to use roman face for Euler's constant e  and the differential d, and concerning spacing before punctuation. I would suggest the following references for mathematics style:
- [Mathematics into Type](http://www.ams.org/arc/styleguide/mit-2.pdf) by Swanson et al (1999), published by the American Mathematical Society (AMS)
- [AMS Style Guide](http://www.ams.org/publications/authors/AMS-StyleGuide-online.pdf), essentially the same guidelines as in *Mathematics into Type* but even more detailed, aimed at copy editors
- Wikipedia guidelines on [Typesetting of mathematical formulae](https://en.wikipedia.org/wiki/Wikipedia:Manual_of_Style/Mathematics#Typesetting_of_mathematical_formulae)

Those all agree that Euler's `e` and integral's `d` are italicised, that a thin space (`\,`) is to be manually inserted before `dx` and that there is no spacing between and equation and punctuation following it. AMS does *not* insert a space before Euler's `e` in `f(x) e^x`, see the examples in Sec 13.11 of the AMS Style Guide.

jessica-mitchell replied:
Fixed, i added only the first refernece in the styleguide as I think that should be enough

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+"see also", "attention", "caution", "danger", "error", "hint", "important",
+"note", "tip", "warning", "admonition"
+
+If you want a custom admonition use
```
This leave me unsure if I could choose any "custom name" myself or are limited to the list given above. But the the "admonition types include" above suggests there are more—where would I find a complete list?

Should we restrict use to certain admonition types in NEST to avoid too much variety?

jessica-mitchell replied:
I think this is the complete list, but I don't think we should use them all,  and perhaps we should specify when to use one or the other. I think right now we have used note, see also, and warning, and possibly important,

jessica-mitchell replied:
Updated text

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+Roles
+~~~~~
+
+See `the roles section in the Sphinx docs <https://www.sphinx-doc.org/en/
```
I wonder if the split between roles and directives is not too technical, especially if it has not been explained before. As a doc writer, I want to put in math, probably mixing inline and displayed math, and then I would look for "how to put in math" and would expect information on inline and displayed math in one place.

I also find the term "Role" in RST-lingo very confusing.

jessica-mitchell replied:
Good point, I'll update this.

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+In-text math
+^^^^^^^^^^^^
+
+For in-text math, use
```
Could point out that full LaTeX formatting can be used.

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+    .. math::
+
+            f(x) = \int_{-\infty}^{\infty} \hat f(\xi)\ e^{2 \pi i x \xi}\,d\xi,
```
Is it possible to enumerate equations and refer to them? Would we want to encourage that? Or discourage?

jessica-mitchell replied:
Yes, you can use the label option (from Sphinx website: When given, it selects an internal label for the equation, by which it can be cross-referenced, and causes an equation number to be issued). 

Being able to cross-reference would be really nice. However, I don't know how the equation numbers are issued over different files. I can look into this

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+.. note::
+
+    The `ref_name` needs to be above the section you want to reference with the
+    following syntax
```
Is the empty line between the ref name and the section header required? If so, point it out.

jessica-mitchell replied:
done

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+.. attention::
+
+  The items in this section are still in development and have not been incorporated into nest:master!
```
In what sense are they still in development? We are using standard Sphinx features here, aren't we? And is the `nest:master` notation clear?

jessica-mitchell replied:
These are all PRs or soon-to-be PRs in different stages of development.

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+  The items in this section are still in development and have not been incorporated into nest:master!
+
+
+To link PyNEST API functions used in-text to the API reference page use the following syntax:
```
`PyNEST API functions` should link to an explanation in the glossary.

jessica-mitchell replied:
Why is that? We don't want to duplicate information that's already in the API docstring, is there additional information we need that would justify a separate glossary entry?

heplesser replied:
I did not mean linking the individual functions to the glossary, I meant that we should have a glossary entry for the term "PyNEST API function" itself: what makes a function a part of the PyNEST API vs a function that is (just) part of the PyNEST code base.

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+    ::
+
+        .. _ref_name:
```
Should we have conventions for reference names? My personal practice in LaTeX is to use a prefix indicating the type of material labeled, eg `fig_`, `eq_`, `sec_`, ... .

jessica-mitchell replied:
This is a good idea, but I don't think we've done this up to this point in the restructured text files.

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+* No formatting such as italics, bold, or underline.
+* Full title of journal
+* Article titles written in sentence case
+* Year follows author(s), in parentheses
```
Move this after information on author names, so it is in the order of appearance in the bib entry.

jessica-mitchell replied:
done

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+  separates volume and page range.
+* Include a linked DOI, if available
+
+.. code-block:: none
```
Why this code block?

jessica-mitchell replied:
I think I wanted to use the `::` but this causes odd highlighting so the code-block (none) removes the highlighting. But I realized I don't actually need it those at all.

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+Headings and subheadings
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Headings and subheadings should describe the purpose of the section.
```
Should this be a bullet or enumerated list?
```suggestion
Headings and subheadings describe the purpose of the section.
```

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+Begin with a descriptive verb or begin with `How to ...`
+
+Headings should explain the section in a short phrase.
```
How does this relate to the first point?

```suggestion
Headings explain the section in a short phrase.
```

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+Avoid section names like `Introduction` or `Part 1`.
+
+One word subheadings are acceptable, if the section is short and the meaning is clear.
```
```suggestion
One-word subheadings are acceptable, if the section is short and the meaning is clear.
```

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+~~~~~~~
+
+Numbers should be spelled out if they begin a sentence. In most cases, however,
+the numeral/ordinal format is preferred.
```
Isn't the general rule that numbers up to twelve are written out?

jessica-mitchell replied:
Yes or 0-9, or 0-10 or 0-100 depending on who you ask :). but there are also exceptions - I'm just looking at the [MS style guide for all their exceptions](https://docs.microsoft.com/en-us/style-guide/numbers)

 But the text could be more precise, and we could specify important exceptions relevant to NEST.

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+Numbers should be spelled out if they begin a sentence. In most cases, however,
+the numeral/ordinal format is preferred.
+
+For additional mathematical notation, use the math role or directive.
```
Add links to math doc.

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+Lists
+~~~~~
+
+Use ordered lists for step-by-step instructions only. Do not have more that 2
```
ordered = numbered lists?

jessica-mitchell replied:
yes, numbered is much more clear than ordered

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+Use ordered lists for step-by-step instructions only. Do not have more that 2
+related actions in one step.
+
+Use bullet lists to improve clarity of long lists (more than 5 items).
```
So what kind of list type to use for up to four items?
```suggestion
Use bullet lists to improve clarity of long lists (more than five items).
```

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+Use bullet lists to improve clarity of long lists (more than 5 items).
+
+If bullet/ordered list text is a complete sentence, use proper punctuation and
```
```suggestion
If the text in a list forms a complete sentence, use proper punctuation and
```

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+If bullet/ordered list text is a complete sentence, use proper punctuation and
+end with period.
+
+If bullet/ordered list text is an incomplete sentence, do not end with period.
```
```suggestion
If the text of a list does not from an incomplete sentence, do not end with period.
```

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+If bullet/ordered list text is an incomplete sentence, do not end with period.
+
+If bullet/ordered list belongs to a sentence, use commas to separate each item
+with the second last time including `and` at end.
```
```suggestion
If a list is part of a sentence, end each item with a comma and the second-last item with ", and". If the sentence ends with the last item on the list, end it with a fullstop, otherwise use the punctuation required to correctly connect to the remainder of the sentence.
```

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+Avoid the pronoun "I".
+
+Avoid gendered terms (NOT police man BUT police officer).
```
```suggestion
Avoid gendered terms (use "police officer" instead of "policeman").
```

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+Avoid gendered terms (NOT police man BUT police officer).
+
+Instead of "guys" or "girls" use inclusive language such as everyone, all,
+members, or folks.
```
```suggestion
Instead of "guys" or "girls" use inclusive language such as "everyone", "all",
"members", or "folks".
```

jessica-mitchell replied:
I know in a previous comment I mentioned the use of quotes can be interpreted as sarcasm, but reading this I find the quotes to be suitable and correct.  I think it's that this sentence is about these words, whereas I was thinking of misusing quotes to emphasize a part of the sentence in way that modifies the tone.

heplesser replied:
I agree: when writing about specific words as words, quotes are in place, as they would be when quoting directly, although that is probably a very rare case in NEST documentation. Should we extend the recommendations on use of quotes correspondingly?

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+members, or folks.
+
+"They" is an acceptable singular third person pronoun
+(see `here <www.merriam-webster.com/dictionary/they>`_).
```
I really wished there was a good term to express third-person singular for those who are neither "he" nor "she", without confusing singular and plural. In Norwegian and Swedish and Danish we have "hen" to complement "han" and "hun", but I guess we just got lucky here with an easy starting point. But for now we should follow the developments in American English as you suggest.

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+Commas
+~~~~~~
+
+Use the oxford comma (apples, bananas, and grapes) for lists. But use a bullet
```
```suggestion
Use the serial comma (apples, bananas, and grapes) for lists. But use a bullet
```

heplesser commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+~~~~~~
+
+Use the oxford comma (apples, bananas, and grapes) for lists. But use a bullet
+list if your list is more than 5 items.
```
Doesn't the question of bullet list vs inline not also depend on the length of the individual items?

jessica-mitchell replied:
Yes, good point. This may be too much of an over-generalization.

jessica-mitchell replied:
reworded

heplesser reviewed (changes requested):
@jessica-mitchell Very nice! I added quite a number of comments in the text.

I have a number of overall questions

- Should we say something about the use (or not) of contractions, e.g., "doesn't", "isn't", etc?
- A policy on punctuation of and around "eg" and "ie"?
- A specific style manual for English, e.g., Strunk & White, APA, ...?


The guide currently does not say anything about the content (sections, ordering) of different types of documentation. I think we have essentially three types of "things" documented:
- PyNEST functions, classes, etc, which should follow Python docstring guidelines. There is a PEP for this, but we may want to repeat the most important parts here, with adjustments for NEST.
- NEST models, where we should make recommendations for what goes where, how much math to include, how to describe model parameters and properties.
- Documentation pages written directly as RST files for overarching documentation.

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+specific and exact terms are needed, which is often. But explanations or
+references should be provided for clarity in introductory texts.
+
+If you have further questions about style not addressed here, see <<STANDARDIZED STYLE GUIDE>>
```
@heplesser  This is an external guide that we could use as a reference; basically the idea is that we only keep a limited set of guidelines or rules and use a comprehensive source, such as the [Microsoft style guide](https://docs.microsoft.com/en-us/style-guide/welcome/), as an authority on questions that arise. For example, if someone wants to know if we should put spaces around an em dash - rather than adding it to our guide, we refer person to reference. I explicitly mention the Microsoft Style Guide because it is and will very likely continue to be well maintained and is a style guide for the computer screen. So the MS guide _or something similar_ may be  a better choice than say, the Chicago Manual or APA style guides.

This is just a suggestion - I saw other projects do this, and thought it could be useful, but it's certainly not a necessity.

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+Writing documentation: a few tips
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Be concise - Use short words and sentences; avoid unnecessary modifiers
```
@heplesser I agree with you; a hyphen is a poor choice here, and I will update this. But I learned a new German word! Gedankenstrich = em dash :)

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+- empathetic - we write thinking of who we are writing for and what their needs
+  are
+- relaxed - we write in natural, conversational way rather than an authoritative
+  way
```
Yes, this is probably good to mention. as fun and tempting as it can be, not everyone knows the same references.

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+~~~~~~~~~~~~~~~
+
+We use double quotes for strings in code, for example
+(nest.Create("iaf_psc_alpha")).
```
Yeah, I wasn't sure about this one as we have a lot of both types. 
Although consistency over all files is nice, an alternative is to say you can choose either single or double quotes, but be consistent within a single document/script. Or something like restructured text documentation files should use double quotes / single quotes.

clinssen replied:
We faced (actually, are still facing) this issue in the NESTML codebase as well, which is a horrible incongruous mix of single and double quotes. We want to standardise on double quotes, as PEP 257 dictates the use of double quotes in triple quoted strings. Global consistency should definitely be a desideratum, but leaves one to wonder how strongly this can be enforced in practice (this will almost definitely be overlooked in PR reviews *n* years down the line).

heplesser replied:
I added this question and the style guide as such to the "in depth" part of the next [ONDVC](https://github.com/nest/nest-simulator/wiki/2021-04-26-Open-NEST-Developer-Video-Conference).

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+We use double quotes for strings in code, for example
+(nest.Create("iaf_psc_alpha")).
+
+Double or single quotes should not be used as an emphasis marker for text.
```
you should not use the "double quote" or 'single quote' to emphasize text. 
If you have something "important" to highlight - the quotes can be read as sarcasm.

jessica-mitchell replied:
improved text

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+\``Double backticks\``
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Use double backticks for
```
Right now there is no real logic; it's mostly based on what we already do in (most/some) cases. But I think this would be a good idea, as long as we consider what the final output will look like and ensure readability is not negatively affected.

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+  - inline code
+  - objects/functions
+  - model names
+  - NEST-specific vocabulary
```
Currently, a catch-all for any other terms that have a NEST specific definition but not included in other categories. Perhaps something like 'events' (although just thinking about this - these terms should be linked to the glossary, and thus will automatically have the format set for restructured text roles . . . but we could adapt the format).

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+Use single backticks for
+
+- Keys
```
a key as in a key, value pair
This should be made more clear

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+\**Boldface\**
+~~~~~~~~~~~~~~
+
+If you want to emphasize a word or phrase in text, you can use **boldface**.
```
Good point, I just have never seen the ** used outside of boldface. 

I have no strong opinions on the use of plain emphasis or basically any of these markups other than I feel italics are less noticeable than bold (or fixed width terms, especially in orange). I also think we should be careful how we use different formatting styles. **We** don't `want to `end up **with** sentences _like this_.

heplesser replied:
The general rule should be to use emphasis sparingly and we should make that clear.

How about the following:
- Plain emphasis when introducing *new terms* in a text, should then probably link to glossary
- Boldface only in **exceptional** cases when overlooking could cause severe problems but the text in question is too short to warrant an admonition box.
- Backticks (single or double, policy to be decided) for class, function, variable names and similar to be entered precisely so in code.

jessica-mitchell replied:
These sound like good ideas to me, but note that if new terms are links to a glossary, they will be formatted as all other links of that role type.

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+==================================
+
+.. contents::
+  :local:
```
the option :local: just needs to be indented, and if there were additional items for the directive, they would all have to have the same indentation. I dont know of a specific rule for how big of an indent it should be.

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+  - inline code
+  - objects/functions
+  - model names
```
I think the formatting markup we want for all these items needs to be discussed. One thing that came up is that we should consider meta-characters (I think is what they're called). For example, the use of the double back slash `\\` --if it's not in double back ticks, it will show up as a single slash. In those cases, we need to use double back ticks or escape the characters, but that might be confusing if looking at the raw rst files.

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+`explained here <https://numpydoc.readthedocs.io/en/latest/format.html>`_.
+
+
+reStructuredText markup
```
I am rearranging the document, and putting the restructured  text header above

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+-------------------
+
+We use specific formatting marks to help readers scan through and find what
+they need, as well as provide references to important terms.
```
Sections have been rearranged so the links to reST docs are further up

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+reStructuredText markup
+-----------------------
+
+Let us take a brief look into the markup of reStructuredText. You can find
```
fixed

jessica-mitchell commented on doc/userdoc/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,566 @@
+
+Let us take a brief look into the markup of reStructuredText. You can find
+a full cheatSheet `here <https://thomas-cokelaer.info/tutorials/sphinx/
+rest_syntax.html>`_.
```
cheat sheet is removed, docutils and sphinx docs used instead

jessica-mitchell commented:
@jougs @heplesser I may have messed up the 'files changed'; I moved the styleguide into the contribute directory, and now Github doesnt follow the history of changes and considers the styleguide a completely new file. From what I quickly googled, following the history after a rename can't be done on Github and is an outstanding issue. 

Should I / Can I  revert back to the old directory so the changes are visible? 

On another note, for discussing this in the ODVC,  I plan to collect all ongoing discussions mentioned above; I can put them all here in a comment, or would it be better to put it in the meeting's wiki?

heplesser commented:
> Should I / Can I revert back to the old directory so the changes are visible?

I think it is most important to check the style guide for internal consistency and the rules it now proposes. I fear no-one really worked much with the old style guide, so changes compared to the old guide do not matter much. A key point here is anyways the introduction of RST, which is quite new to NEST. So I would leave it as it is now.
> 
> On another note, for discussing this in the ODVC, I plan to collect all ongoing discussions mentioned above; I can put them all here in a comment, or would it be better to put it in the meeting's wiki?

To keep all discussions in one place and most visible, I'd put the summary here and then add a link from the Wiki.

jessica-mitchell commented:
> > 
> 
> I think it is most important to check the style guide for internal consistency and the rules it now proposes. I fear no-one really worked much with the old style guide, so changes compared to the old guide do not matter much. A key point here is anyways the introduction of RST, which is quite new to NEST. So I would leave it as it is now.

Sorry I don't think I was clear as to what I meant -  Github considers the entire styleguide.rst a new file becasue I moved it into a different directory - so you cannot see the diff of what I've changed based on your suggestions.

clinssen commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,623 @@
+"Double quotes"
+~~~~~~~~~~~~~~~
+
+We use double quotes for strings in code, for example
```
```suggestion
We use double quotes for strings in code, for example ``nest.Create("iaf_psc_alpha")``. This applies to Python as well as C++ code. This rule is based on PEP 257, which (only) dictates the use of double quotes in triple quoted strings; for consistency, double quotes are used throughout the codebase.
```

jessica-mitchell replied:
done

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,623 @@
+~~~~~~~~~~~~~~~~~~~~~~
+
+In PyNEST code, we follow the rules for NumPy style docstrings as
+`explained here <https://numpydoc.readthedocs.io/en/latest/format.html>`_.
```
Should we mention explicitly here that this implies that in PyNEST documentation, one shall only use the sections listed in the NumPy doc guide, in that order, and not use any admonitions? Or at least a sentence explicitly asking documenters to carefully read the NumPy Guide before working on PyNEST docstrings?

jessica-mitchell replied:
Added text

heplesser commented on doc/userdoc/contribute/index.rst:
```diff
@@ -9,7 +9,8 @@ Please familiarize yourself with our:
 * :doc:`NEST Git workflow <development_workflow>`
 * :doc:`C++ coding style guidelines <coding_guidelines_cpp>`
-* :doc:`Developing NEST with IDEs <nest_with_ides>`
+* :doc:`Documentation style guide <templates_stylguides/styleguide>`
```
```suggestion
* :doc:`Documentation style guide <templates_styleguides/styleguide>`
```

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+The NEST documentation style guide
```
In my view, templates and style guides are very different types of documents. I would not combine them in one folder.

jessica-mitchell replied:
yes, I tend to agree I will separate them

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+==================================
+
+.. contents:: On this page, you'll find
+   :local:
```
This list of contents is very long and unstructured. Can we organise it differently?

jessica-mitchell replied:
I moved content around and changed the headings to try to better separate the sections. It should be checked if it makes sense. I also limited the contents on the right side bar to a depth of 1 level

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+You will also need to know how to build the documentation locally on your machine: See
+:doc:`../../documentation_workflow/user_documentation_workflow`.
+
+For additional information, including the Git workflow and making a pull request see :doc:`../index`
```
```suggestion
For additional information, including the Git workflow and making a pull request see :doc:`../index`.
```

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+A few tips on writing documentation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Be concise: Use short words and sentences; avoid unnecessary modifiers
```
This is not rendered as a list, just a sequence of paragraphs. Several items on the list miss the period at the end.

jessica-mitchell replied:
All of these are rendering as lists for me.

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+Our voice is
+
+- empathetic: we write thinking of who we are writing for and what their needs
```
Not rendered as list. Periods missing at end.

jessica-mitchell replied:
Periods added

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+  way
+- factual: we write in a neutral style avoiding irony, humor, or cultural references
+
+Use 'you' to indicate the reader and 'we' to indicate yourself (the writer and
```
```suggestion
Use "you" to indicate the reader and "we" to indicate yourself (the writer and
```

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+We follow spelling and grammar rules of `American English <https://www.merriam-webster.com/>`_.
+
+The markup language we use is :ref:`reStructuredText <sec-reST_style>`. This includes documentation
+blocks in files written in Python and cpp.
```
```suggestion
blocks in files written in Python and C++.
```

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+We do, however, use terminology appropriate to NEST and neuroscience when
+specific and exact terms are needed, which is often. But explanations or
+references should be provided for clarity in introductory texts and the glossary.
```
Add link to glossary.

jessica-mitchell replied:
link added

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
````diff
@@ -0,0 +1,608 @@
+In addition to directives, reStructuredText has roles, which insert semantic markup into documents.
+Roles look like this ``:role-name:`content```.
+
+We will only cover a few examples here. You can find more information in the following links.
````
```suggestion
We will only cover a few examples here. You can find more information in the following links:
```

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+We will only cover a few examples here. You can find more information in the following links.
+
+
+* `reStructuredText User Documentation <https://docutils.sourceforge.io/rst.html#id24>`_
```
Not rendered as list.

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+    .. code-block:: cpp
+
+        int main()
+            {
+              cout << "Hello, World!"; \\
+              return 0;
+            }
```
Use NEST C++ formatting rules:
```suggestion
        int main()
        {
          std::cout << "Hello, World!";
          return 0;
        }
```

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+    .. code-block:: cpp
+
+        int main()
+            {
+             cout << "Hello, World!";
+             return 0;
+            }
```
```suggestion
        int main()
        {
          std::cout << "Hello, World!";
          return 0;
        }
```

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+For Python examples that show input and output use the following syntax::
+
+   >>> input begins with 3, right-angled brackets
+   output is directly below input without any brackets. A blank line must end the example.
```
```suggestion
   Output is directly below input without any brackets.
   A blank line must end the example.
```

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+Use the admonition
+
+* "See also" to reference internal or external links (only in cases where the reference should stand out),
```
Not rendered as list

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+   :doc:`path/to/file`
+
+In this case, the link caption will be the title of the given document.
```
What is a link caption?

jessica-mitchell replied:
modified wording and added an additional file to better exemplify how these references work

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+   :doc:`Top header of file`
+
+   :doc:`custom label <file>`
```
I do not understand this example when looking at it in rendered form.

jessica-mitchell replied:
added an additional file to better exemplify how these references work

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+To reference figures or arbitrary places in a file, you must include a custom
+label in the reference for it to work::
+
+    :ref:`custom label <my-arbitrary-place-label>`
```
I don not understand the custom label here. Could you give a proper example?

jessica-mitchell replied:
added

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+.. attention::
+
+  The items in this section are still in development and have not been incorporated into nest:master!
```
I am still not sure what you mean here, especially with the "have not been incorporated into master". Do you want to say that not all documentation in NEST code currently follows these guidelines to the dot, or do you mean that the rules are still subject to change? If the former, shouldn't be put that as a general disclaimer at the top of this document and point the users explicitly to a small number of documents/PyNEST docstrings/... that are fully following these rules already, so they have them as worked examples; and explicitly point out that all else is in some stage of transition and *should not* be consulted as reference for how to do things?

jessica-mitchell replied:
This attention admonition is meant to be removed before this PR is merged - it is meant for anyone looking at the document while in review, and wondering why they have never seen these types of links in the documentation (I see how that is not obvious).
I thought it was important to mention intersphinx, hoverxtooltip and the function links in case there were discussions revolving around these topics with respect to to style guide. And I wasn't sure if they'd be ready for review before this PR was merged. Anything not in master mentioned here, should be removed before this PR is merged imo.

jessica-mitchell replied:
removed admonition

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+Rendered as
+
+   :py:func:`.Create`
```
Does one really need the dot? When do I need the dot?

jessica-mitchell replied:
Yes the dot is necessary to generate the link. I'll add an explanation

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
````diff
@@ -0,0 +1,608 @@
+
+   ``:py:func:`~nest.lib.hl_api_types.SynapseCollection.funcname```
+
+   ``:py:func:`~nest.lib.hl_api_types.NodeCollection.funcname```
````
Could you explain why?

jessica-mitchell replied:
It turns out I was mistaken as I  assumed get() and set() were functions but they are actually methods. Still, because they are both used in two classes, they have slightly special markup. I added text explaining this.

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+Links to glossary
+^^^^^^^^^^^^^^^^^
+
+To link terms to the glossary page use the HoverXTooltip role :hxt_ref: from Mahdi Enan (INM-6)::
```
Provide link to extension. Most external readers will not know INM-6.

jessica-mitchell replied:
link added

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+Links to certain external projects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+With the Sphinx extension 'intersphinx', projects that also use Sphinx/ReadtheDocs can be referenced the same way as
```
Are hyphens around intersphinx correct? Maybe rather make it a link?

jessica-mitchell replied:
link added

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+Please ensure your reference follows the following guidelines:
+
+*  Do not add formatting markup such as italics, bold, or underline.
```
Not rendered as list.

jessica-mitchell replied:
I'm not sure why this is - the bullets render correctly for me.

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+In general, we try to follow the pattern of
+
+* First heading: ===
```
Note rendered as list.

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+Use double backticks for
+
+
+  - inline code
```
Not rendered as list.

heplesser commented on doc/userdoc/contribute/templates_styleguides/styleguide.rst:
```diff
@@ -0,0 +1,608 @@
+
+Use single backticks for
+
+- Dictionary keys (if no value is provided)
```
Note rendered as list.

heplesser reviewed (changes requested):
@jessica-mitchell Thank you for the updates, it looks pretty good now! One systematic problem seems to be that none of the bullet lists got rendered as proper lists (with bullets), just sequences of paragraphs.

heplesser commented:
@jessica-mitchell Thanks for all the fixes! But I still see no bullet points in the rendered lists. Looking at the HTML code reveals that the lists are properly encoded as lists, but with `<ul class="simple">` and apparently
```
dl, ol, ul {
    margin: 0;
    padding: 0;
    list-style: none;
    list-style-image:none
}
```
in `theme.css`. I assume the zeros and nones here are the problem. It is the same with Safari and Firefox. Do you have any ideas?

jessica-mitchell commented:
Points still in need of discussion:

- Should we say something about the use (or not) of contractions, e.g., "doesn't", "isn't", etc?
- A policy on punctuation of and around "eg" and "ie"?
- Conventions for reference labels (e.g., use sec_, fig_ eq_ at the beginning of each reference label to denote the type of reference)
- numerate equations and [refer to them](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#math)? 
- Formatting: backticks, emphasis (bold, italic)
    -  items which have markup with  :role:\`item\` 
        - glossary items 
        - functions, classes etc. 
        - documents, sections within documents 
        - math equations
        - [Sphinx has other possible roles](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html)
    - What formatting should the following items have?
        - in line code  - dbl back tick
        - paths/filenames - dbl back tick
        - commands - dbl back tick
        - model names - dbl back tick
        - parameters
        - variables (x)
        - values
        - variables with assigned values (x = 1)
        - values with symbols ( 5 mV)
        - types (bool, int)
        - meta characters (backslashes)
        - special characters (ampersands, math symbols)
        - words we want emphasized in text - bold / italic?

clinssen commented on doc/userdoc/contribute/styleguide/sample_doc.rst:
```diff
@@ -0,0 +1,18 @@
+
+.. math::
+
+    e = mc^2
```
```suggestion
    E = mc^2
```

jougs commented:
@jessica-mitchell: would it be possible to post a link to a rendered version? Thanks!

clinssen commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,649 @@
+*  Add issue in parentheses after volume (optional).
+*  Include a linked DOI, if available.
+
+.. code-block:: none
```
Is this line supposed to be here? References below are not rendered this way.

jessica-mitchell replied:
Yes, I use the none code-block for the example references so I don't get the highlighting that `::` produces. It's not meant to show you how to create references - it's just in the raw reStructuredText file to remove syntax highlighting. See below for what it looks like with the double colon:

![image](https://user-images.githubusercontent.com/3313313/117856270-b1e97f00-b28b-11eb-8af7-0000f4d67948.png)

clinssen commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,649 @@
+
+   reStructuredText is sensitive to indentation and new lines.
+
+   * Directives, headings, labels, and tables should be separated from other text by a new line, excluding directive options.
```
```suggestion
   * Directives, headings, labels, and tables should be separated from other text by a newline, excluding directive options.
```

clinssen commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,649 @@
+If the abbreviation/acronym is well known (e.g., HTML), you do not need to spell
+it out.
+
+Use "e.g.," and "i.e.," only in parentheses and figure and table captions; otherwise, use "for example," and "that is,."
```
```suggestion
Use "e.g." and "i.e." only in parentheses and figure and table captions; otherwise, use "for example" and "that is."
```

clinssen commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,649 @@
+
+If the text of a list forms an incomplete sentence, do not end with period.
+
+If the entire bullet/numbered list belongs to a sentence, end each item with a comma and the second-last item with ", and."
```
```suggestion
If the entire bullet/numbered list belongs to a sentence, end each item with a comma and the second-last item with ", and".
```

jessica-mitchell commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,649 @@
+
+.. note::
+
+   Note the object name is prefixed with a dot.
```
fixed

jessica-mitchell commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,649 @@
+
+.. note::
+
+   reStructuredText is sensitive to indentation and new lines.
```
fixed

jessica-mitchell commented:
@clinssen most of your suggestion I added (see exceptions in comments above), including references at the bottom of the page. Thanks!

clinssen commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,649 @@
+~~~~~~~~~~~~~~~~
+
+To link terms to the glossary page, use the HoverXTooltip role :hxt_ref:. The source code for the extension can be
+`found here <https://github.com/INM-6/HoverXTooltip>`_. ::
```
It could be that the repository is marked as private; try opening it in an incognito browser tab.

jessica-mitchell replied:
Oh right, I forgot about that, thanks for spotting. I am actually going to remove this section because the hoverxtooltip is not yet implemented.

heplesser commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,634 @@
+
+If you want a custom admonition, use::
+
+   .. admonition:: custom label
```
```suggestion
   .. admonition:: Custom label
```

heplesser commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,634 @@
+
+   .. admonition:: custom label
+
+         Here is some text
```
For some reason, the `is` here is rendered boldface. Do you understand why and can it be avoided?

jessica-mitchell replied:
This is because of syntax highlighting when using  the `::`, I changed it to a code-block with no highlighting

heplesser commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,634 @@
+Rendered as
+
+
+   .. admonition:: custom label
```
```suggestion
   .. admonition:: Custom label
```

heplesser commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,634 @@
+For example:
+
+    The following example is based on Smith [1]_.
+    Sanders et al. [2]_ contains a technically detailed example.
```
I am a bit confused here. Is this meant as an example of how to enter references in RST? Why are the references then rendered orange in Readthedocs? On the other hand, if this is to show the end result, why are the underscores shown in the rendered version?

jessica-mitchell replied:
Yah, I see the confusion. I modified the text to make it a rendered example

heplesser commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,634 @@
+
+Prefer the active voice, where the subject acts on object.
+
+If the object needs emphasis rather than subject, use passive voice. But ensure
```
```suggestion
If the object needs emphasis rather than the subject, use passive voice. But ensure
```

heplesser commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,634 @@
+The language we use
+-------------------
+
+We follow spelling and grammar rules of `American English <https://www.merriam-webster.com/>`_.
```
From here on through a several of the following sections, we have single-sentence paragraphs, many quite short. Would it make more sense and be more pleasing to the reader's eye to turn this all into bullet lists?

jessica-mitchell replied:
I updated all the single sentences so they are bullet lists, I think this looks better thanks

heplesser reviewed (changes requested):
@jessica-mitchell Just a few more details ;).

jessica-mitchell commented:
@heplesser I applied your suggestions, thanks!

heplesser reviewed (approved):
@jessica-mitchell Great work!

heplesser commented:
@terhorstd @jougs You are either set up as reviewers or have commented earlier on the style guide. Do you still want to review/comment or can we go ahead and merge this one with two approvals in place?

terhorstd commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,654 @@
+
+* Use the serial comma (apples, bananas, and grapes) for lists.
+
+* Use the comma as separator for thousands (37,000).
+
```
This is a duplicate of the information given in "Numbers" section. Maybe add a link there, because it holds more information.

terhorstd commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,654 @@
+
+The ``:ref:`` role requires a reference label that looks like this ``.. _type_ref-label:``.
+
+::
+
+   .. _sec_my-ref-label:
+
+   Section to cross-reference
+   --------------------------
+
+   Some content in this section.
+
```
This seems to render the "in" as code. Should this be a `.. code-block:: text` or `.. code-block:: null` ?

jougs replied:
Fixed by my recent suggestion.

terhorstd reviewed (approved):
Just two minor remarks. Since those are not critical we could also add them incrementally. Add them if possible. 
:+1:

jougs commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,654 @@
+   Output is directly below input without any brackets.
+   A blank line must end the example.
+
+For in-text code use the role :code: or double back ticks::
```
```suggestion
For in-text code use the role :code: or double backticks::
```

jougs commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,654 @@
+  `explained in the NumPy style guide <https://numpydoc.readthedocs.io/en/latest/format.html>`_.
+
+* However, we use different formatting marks than what is stated in their guide. See section on
+  :ref:`double back ticks <sec_dbltick>`.
```
```suggestion
  :ref:`double backticks <sec_dbltick>`.
```

jougs commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,654 @@
+
+
+\**Strong emphasis\** vs \*emphasis\*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

jougs commented on doc/userdoc/contribute/styleguide/styleguide.rst:
```diff
@@ -0,0 +1,654 @@
+
+* `American English dictionary <https://www.merriam-webster.com/>`_
+
+* `the Microsoft style guide <https://docs.microsoft.com/en-us/style-guide/welcome/>`_
```
```suggestion
* `The Microsoft style guide <https://docs.microsoft.com/en-us/style-guide/welcome/>`_
```

jougs reviewed (approved):
Beautiful! I have added some minor suggestions.
